Add Contract Deployment guide (#840)
lukso-network/docs: docs/learn/dapp-developer/deploy-contracts.md

---
sidebar_label: '🏭 Deploy contracts'
sidebar_position: 10
description: Learn how to deploy a contract from your Universal Profile.
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Deploy Contracts

<div style={{textAlign: 'center', color: 'grey'}}>
  <img
    src={require('./img/deploy-contract.png').default}
    alt="Deploying a contract from the Universal Profile using the Browser Extension"
    width="400"
  />
<br/>
<i>Deploying a contract from a Universal Profile using the Browser Extension</i>
<br /><br />
</div>

## Setup

```shell

> npm install ethers

Reviewer: Is this using v5 or v6? Maybe good to precise in case?

```

## Deploy a contract from a Universal Profile

The Universal Profile browser extension will magically wrap all the calls internally so you don't have to worry about crafting custom transactions. Simply use [`eth_sendTransaction`](https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_sendtransaction) as you always did while working with EOA.

<!-- prettier-ignore-start -->

```js
import { ethers } from 'ethers';

const provider = new ethers.providers.Web3Provider(window.lukso);

await provider.send("eth_requestAccounts", []);

const signer = provider.getSigner();
const account = await signer.getAddress();

// Send transaction
const tx = await signer.sendTransaction({
    from: account,                        // The Universal Profile address
    value: 0                              // optional, in case constructor is payable
    data: "0x0x608060405234801561001.."   // Full bytecode of the contract to deploy + constructor args
});

const receipt = await tx.wait();
```

> You can utilize the ethers.js contract factory to deploy a contract by supplying its ABI and bytecode. For instance, if you're deploying an LSP7Mintable contract, you can obtain its ABI and bytecode from the `@lukso/lsp-smart-contracts` package.

Reviewer: Suggest adding links to:
- ethers contract factory
- LSP7Mintable ABI

```suggestion
You can use the [contract factory from ethers.js](./link-to-relevant/section/on-ethers-docs) to deploy a contract by supplying its ABI and bytecode. For instance, if you're deploying an [LSP7Mintable](../../contracts/internal-link/to-LSP7Mintable-abi) contract, you can obtain its ABI and bytecode from the `@lukso/lsp-smart-contracts` package.
```

> 

Reviewer: Not sure why there is so many line breaks. Is it intended?

```shell
npm install @lukso/lsp-smart-contracts
```

<!-- prettier-ignore-start -->

```js

import LSP7Mintable from "@lukso/lsp-smart-contracts/artifacts/LSP7Mintable.json"

// Prepare the factory with the abi and bytecode
const lsp7Factory = new ethers.ContractFactory(
      LSP7Mintable.abi,
      LSP7Mintable.bytecode,
      signer
    );

// send the transaction of deployment
const transaction = await lsp7Factory.deploy(
    'Yamen token',
    'no desc',
    "0x0000000000000000000000000000000000000011",
    0,
    false
  );
          
const receipt = await transaction.deployTransaction.wait();
```
<!-- prettier-ignore-end -->

> ## Retreiving Contract Deployment Information

Reviewer: ```suggestion
## Retrieving Contract Deployment Information
```

> There will be no `contractAddress` field in the transaction receipt unlike normal transactions happening with Metamask, as the browser extension wraps the transaction of deployment with a normal contract execution transaction. The contract address can be calculated using the **nonce** and the **address** of the profile, or retreived using the logs:

Reviewer: ```suggestion
There will be no `contractAddress` field in the transaction `receipt` unlike normal transactions happening with Metamask. This is because the UP browser extension wraps the deployment transaction within a normal contract execution transaction. The contract address can be calculated using the **nonce** and the **address** of the profile or retrieved using the logs:
```

Reviewer: Also suggest to mention that for the code snippet below, you are using the 2nd method using the logs, so that developers reading have the context in mind before reading the code snippet that follows :)

---
sidebar_label: '🏭 Deploy contracts'
sidebar_position: 10
description: Learn how to deploy a contract from your Universal Profile.
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Deploy Contracts

<div style={{textAlign: 'center', color: 'grey'}}>
  <img
    src={require('./img/deploy-contract.png').default}
    alt="Deploying a contract from the Universal Profile using the Browser Extension"
    width="400"
  />
<br/>
<i>Deploying a contract from a Universal Profile using the Browser Extension</i>
<br /><br />
</div>

## Setup

<Tabs groupId="web3-lib">

  <TabItem value="ethersjs" label="ethers.js">

```shell
npm install ethers
```

  </TabItem>

> </Tabs>

Reviewer 2: why use tab item if there is only one tab??

If one tab then just remov the tab and put the content directly
